cmd/utils/executor.go

```diff
@@ -0,0 +1,243 @@
+	"golang.org/x/term"
+)
+
+type ManifestExecutor interface {
+	Execute(command string, env []string) error
+}
+
+type Executor struct {
+	outputMode string
+	displayer  executorDisplayer
+}
+
+type executorDisplayer interface {
+	display(scanner *bufio.Scanner, command string, sb *screenbuf.ScreenBuf)
+	startCommand(cmd *exec.Cmd) (io.Reader, error)
+}
+
+type ttyExecutorDisplayer struct{}
+type plainExecutorDisplayer struct{}
+type jsonExecutorDisplayer struct{}
+
+type jsonMessage struct {
+	Level     string `json:"level"`
+	Stage     string `json:"stage"`
+	Message   string `json:"message"`
+	Timestamp int64  `json:"timestamp"`
+}
+
+// NewExecutor returns a new executor
+func NewExecutor(output string) *Executor {
+	if output == "tty" && !isSupportForTTY() {
+		output = "plain"
+	}
+	var displayer executorDisplayer
+	switch output {
+	case "tty":
+		displayer = ttyExecutorDisplayer{}
+	case "plain":
+		displayer = plainExecutorDisplayer{}
+	case "json":
+		displayer = jsonExecutorDisplayer{}
+	}
+	return &Executor{
+		outputMode: output,
+		displayer:  displayer,
+	}
+}
+
+// Execute executes the specified command adding `env` to the execution environment
+func (e *Executor) Execute(command string, env []string) error {
+
+	cmd := exec.Command("bash", "-c", command)
+	cmd.Env = append(os.Environ(), env...)
+
+	reader, err := e.displayer.startCommand(cmd)
+	if err != nil {
+		return err
+	}
+
+	scanner := bufio.NewScanner(reader)
+
+	sb := screenbuf.New(os.Stdout)
+	go e.displayer.display(scanner, command, sb)
+
+	err = cmd.Wait()
+	if e.outputMode == "tty" {
+		collapseTTY(command, err, sb)
+	}
+	return err
+}
+
+func startCommand(cmd *exec.Cmd) (io.Reader, error) {
+	reader, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+	return reader, nil
+}
+
+func (plainExecutorDisplayer) startCommand(cmd *exec.Cmd) (io.Reader, error) {
+	return startCommand(cmd)
+}
+
+func (plainExecutorDisplayer) display(scanner *bufio.Scanner, _ string, _ *screenbuf.ScreenBuf) {
+	for scanner.Scan() {
+		line := scanner.Text()
+		fmt.Println(line)
+	}
+}
+
+func (jsonExecutorDisplayer) startCommand(cmd *exec.Cmd) (io.Reader, error) {
+	return startCommand(cmd)
+}
+
+func (jsonExecutorDisplayer) display(scanner *bufio.Scanner, command string, _ *screenbuf.ScreenBuf) {
+	for scanner.Scan() {
+		line := scanner.Text()
+		level := "info"
+		if isErrorLine(line) {
+			level = "error"
+		}
+		messageStruct := jsonMessage{
+			Level:     level,
+			Message:   line,
+			Stage:     command,
+			Timestamp: time.Now().Unix(),
+		}
+		message, _ := json.Marshal(messageStruct)
+		fmt.Println(string(message))
+	}
+}
+
+func (ttyExecutorDisplayer) display(scanner *bufio.Scanner, command string, sb *screenbuf.ScreenBuf) {
+	queue := []string{}
+	for scanner.Scan() {
+		commandLine := renderCommand(command)
+		sb.Write(commandLine)
+		line := scanner.Text()
+		if len(queue) == 10 {
```

Comment:
can we do this value configurable via a command argument and envvar? also, is i possible to default to 25?